### PR TITLE
Added MSAA to GLES2 backend

### LINKED ISSUE #25843: Anti aliasing in 3.1 GLES2 Renderer
It looks like there is no way to get 3D anti aliasing in the new GLES2 renderer. The MSAA option is still present, but it does nothing. I'm currently evaluating the GLES2 renderer for my game but no AA is a deal breaker.

### PR DESCRIPTION
**Pull request by clayjohn**
As discussed with @reduz, this PR checks to see if the multisample FBO extension is available on the hardware, if so it renders to a multisample renderbuffer and blits back to the normal framebuffer at the end of frame or when the screen is being read from. 

Fixes #25843

### CODE REVIEW

**Comment by akien-mga** [comment]
Note for release managers: if/when merged, let it simmer in the master pot for a while for potential fixups, but otherwise it should be a nice candidate for cherry-picking in 3.1.2.

**Review by reduz — APPROVED** [review]

**Comment by clayjohn** [comment]
Changes to fix the travis ci

**Comment by clayjohn** [comment]
@akien-mga when you get a chance I could use a hand with the travis-ci errors. I have resolved most, the last one is only for android and iphone. Neither one properly recognizes the OpenGL-specific constants. They are used the same as in GLES3, so they should be defined somewhere, but I can't figure out what travis needs in order to recognize that.

**Comment by akien-mga** [comment]
@clayjohn The problem is that those methods and constants are not defined in OpenGL ES 2.0, but they are in OpenGL 2.1. From a quick look, there doesn't seem to be widely supported extensions to bring those to OpenGL ES 2.0.

So this code should be enclosed in `#ifdef GLES_OVER_GL` to be OpenGL-only, or an alternative implementation will be needed to be compatible with the OpenGL ES 2.0 spec.

**Comment by akien-mga** [comment]
> So this code should be enclosed in #ifdef GLES_OVER_GL to be OpenGL-only, or an alternative implementation will be needed to be compatible with the OpenGL ES 2.0 spec.

I guess we need to make sure to limit the API calls to things enabled by https://www.khronos.org/registry/OpenGL/extensions/EXT/EXT_framebuffer_multisample.txt (for OpenGL 1.5+ so `GLES_OVER_GL` for us) and https://www.khronos.org/registry/OpenGL/extensions/EXT/EXT_multisampled_render_to_texture.txt (for OpenGL ES 2.0).

BTW, useful reference to find matching extensions: https://chromium.googlesource.com/angle/angle/+/master/src/libANGLE/renderer/gl/FeatureSupportGL.md

**Comment by akien-mga** [comment]
Also:

> If OpenGL ES 3.0 or later is not supported, ignore all references
        to DRAW_FRAMEBUFFER and READ_FRAMEBUFFER.

in https://www.khronos.org/registry/OpenGL/extensions/EXT/EXT_multisampled_render_to_texture.txt

**Comment by akien-mga** [comment]
So there are several things to do for OpenGL ES 2.0 support:
- Identify and replace the API calls which *can't* be supported, like `glReadBuffer`, `glBlitFramebuffer`, `GL_READ_FRAMEBUFFER` and `GL_DRAW_FRAMEBUFFER`
- Then some of the remaining calls likely need `EXT` appended as they're neither in the OpenGL 2.1 nor OpenGL ES 2.0 specifications, but rely on the two extensions mentioned above and checked against in this patch.
- Then I guess some magic will be needed so that we can make those `EXT` calls conditionally point to the right GLAPIENTRY or to void, based on whether the matching extension is defined.

**Comment by clayjohn** [comment]
@akien-mga Im gonna need some help figuring out the GLAPIENTRY stuff. I thought I could get it to work by including <GLES2/glext.h> because that is where the definitions of the functions are stored, but it didn't work. reduz suggested just defining the functions in the file. That may be our best option. But I'm not sure how that would work while still keeping things nice and clean.

I added stuff to the glad file for 2.1, this should make things work without the EXT call. (I'm just mirroring how it is already done in rasterizer_gles2.cpp)

**Comment by akien-mga** [comment]
@clayjohn The errors are:
```
drivers/gles2/rasterizer_storage_gles2.cpp:4689:3: error: use of undeclared
      identifier 'glRenderbufferStorageMultisample'
                glRenderbufferStorageMultisample(GL_RENDERBUFFER, msaa, ...
                ^
drivers/gles2/rasterizer_storage_gles2.cpp:4711:3: error: use of undeclared
      identifier 'glFramebufferTexture2DMultisample'
                glFramebufferTexture2DMultisample(GL_FRAMEBUFFER, GL_COL...
                ^
```
But as I mentioned above, and as per the extension, you should use `glRenderbufferStorageMultisampleEXT` and `glFramebufferTexture2DMultisampleEXT`, both of which are properly defined in `GLES2/gl2ext.h` (at least in the Android NDK).

File: drivers/gles2/rasterizer_gles2.cpp
Comment (akien-mga) [inline drivers/gles2/rasterizer_gles2.cpp]: Trailing space.

**Comment by akien-mga** [comment]
I tried this fixup locally, and that fixes the build for Android, but then there's a link error for those EXT symbols: https://hastebin.com/nazajegiti

**Comment by clayjohn** [comment]
@akien-mga I added es2 to glad.h and then included it for iphone and android. It builds successfully for android now, but I am unable to test iphone. I had to change an SCsub file in order to get scons to compile glad.h when using android or iphone.

**Comment by clayjohn** [comment]
@akien-mga Just pushed another version. Now im loading the needed functions for ios and android manually. Turns out ios also has a different way of handling multisampled buffers.

Also, I can't build for ios so I'm pretty much just relying on travis-ci to let me know if I made a mistake here...

**Comment by clayjohn** [comment]
@godotengine/ios Can I get someone from the ios team to test this on an ios device? 

@akien-mga This is finally passing travis-ci!!

@reduz Can you review this again as I made a few changes to get it working with ios and android?

**Comment by akien-mga** [comment]
As discussed on IRC, there are some concerns about how this will work on real GLES2-only devices (was only tested on GLES3-capable devices) and especially on iOS.

Since we don't have a battery of devices nor QA testers available to check it before merging, we'll go with our usual workflow: test in production. Please give the feature a try in the master branch and report any issue you encounter.

To release maintainers: this means of course that it shouldn't be cherry-picked for the 3.1 branch right away. We should wait until we're confident that tests in the master branch confirm that the feature works well everywhere.

**Comment by akien-mga** [comment]
Thanks a lot!

**Comment by akien-mga** [comment]
Note to release maintainers: if we decide to cherry-pick this, there are quite a few follow-up PRs to include (search "MSAA" in merged PRs). Overall probably not worth the trouble given that the 3.2 release is imminent and includes a lot of GLES2 fixes.

**Pull request merged**
